---
title: "Intro"
author: "XQS"
date: "March 1, 2021"
output: html_document
runtime: shiny
---

## **Introduction**

---

This is a web-based interactive application for further analysis of single cell data after Seurat analysis.

You can freely choose the analysis method and set the values of the parameters. After the analysis, you can also save the result data for drawing or some personalized analysis.

It is meant to provide an intuitive interface for researchers to easily upload, analyze, visualize, and explore single-cell RNA-seq data interactively with no prior programming knowledge in R.


## **Input Data Types**

---

This application accepts the following types of input data:

### 1. Example data (Demo):

- For demo purposes, you can select "Example data"

- That will automatically load the Seurat object data

### 2. Upload your own data:

- You can choose local file upload
    * 'Data-rds':The input data is a Seurat object
    * 'Data-counts':The input data is the counts matrix, the rows are genes, and the columns are samples. Support multiple formats:csv, rds, and h5
    * 'Data-10X':The input is a folder

## **Run Results**
---
### 1. Data Output
There will be plenty of output information from major steps, some of which will be displayed downloadable.

### 2. Visualization
Various forms of visualizations are included:

* **GSVA**
    * Heatmap
* **Monocle**
    *  cellTrajectory_pse
    *  cellTrajectory_sta
    *  cellTrajectory_seu
* **Scenic**
    *  
    *  
* **CellphoneDB**
    * Bubble map 
    * Heatmap
* **Cor**
    * Heatmap 
    
---

